### ISSUE TITLE
Download a GIF and store the image, SD will save it as a PNG to disk.

### ISSUE THREAD

**Issue by gonghonglou** [SDWebImage/SDWebImage#2952]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | 13.0
 SDWebImage Version      | 5.5.2 / 4.4.8
 Integration Method      | cocoapods
 Xcode Version           | Xcode 11
 Repro rate              | all the time (100%)
 Repro with our demo prj | 
 Demo project link       | 

### Issue Description and Steps

In 4.4.8 version, the image will became a static image form disk.
In 5.5.2 version, even the gif is work from disk, because SD save the image as a UIAnimatedImage, but the format from NSData is also PNG(0x89), not GIF(0x47).

Example code:
```
NSURL *url = [NSURL URLWithString:@"http://img.xxx.gif"];
[[SDWebImageDownloader sharedDownloader] downloadImageWithURL:url options:SDWebImageDownloaderLowPriority progress:^(NSInteger receivedSize, NSInteger expectedSize, NSURL * _Nullable targetURL) {
} completed:^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, BOOL finished) {
    
    [[SDImageCache sharedImageCache] storeImage:image forKey:url.absoluteString completion:nil];
}];
```

Even i can use this method `storeImage:imageData:forKey:toDisk:completion:` to solve the question, but, can SDImageCache change store method on [L192-L197](https://github.com/SDWebImage/SDWebImage/blob/master/SDWebImage/Core/SDImageCache.m#L192-L197) to this:

```
SDImageFormat format;
if (image.sd_imageFormat != SDImageFormatUndefined) {
    format = image.sd_imageFormat;
} else if ([SDImageCoderHelper CGImageContainsAlpha:image.CGImage]) {
    format = SDImageFormatPNG;
} else {
    format = SDImageFormatJPEG;
}
```

**Comment by dreampiggy** [comment]
Don't you see there is a `imageData` arg in `SDImageCache` or `SDImageCacheProtocol` APIs ?

If you give an `UIImage`, there are no any `format` information, so we use PNG or JPEG.
If you need to keep original image format, encode it (You can use `SDImageCoder`, or convienent `UIImage+MultipleFormat` API) and pass `imageData` arg.

Note the default `SDWebImageManager` always keep the original image data, not just PNG/JPEG.

Read documentation here: https://sdwebimage.github.io/SDWebImage/Classes/SDImageCache.html#/c:objc(cs)SDImageCache(im)storeImage:imageData:forKey:toDisk:completion:

**Comment by dreampiggy** [comment]
The code:

```objective-c
                NSData *data = imageData;
                if (!data && image) {
                    // If we do not have any data to detect image format, check whether it contains alpha channel to use PNG or JPEG format
                    SDImageFormat format;
                    if ([SDImageCoderHelper CGImageContainsAlpha:image.CGImage]) {
                        format = SDImageFormatPNG;
                    } else {
                        format = SDImageFormatJPEG;
                    }
                    data = [[SDImageCodersManager sharedManager] encodedDataWithImage:image format:format options:nil];
                }
```

Only encode the UIImage into PNG/JPEG when there are no `imageData`.

**Comment by dreampiggy** [comment]
If you want to change the behavior into:

```
if ([image conformsToProtocol:@protocol(SDAnimatedImage)]) {
    data = [image animatedImageData];
}
// ...
if (image.sd_isAnimated) {
    // encode Animation
    imageFormat = SDImageFormatGIF; // or SDImageFormatAPNG ? It's better
}
```

This may introduce new behavior, both in 4.x and 5.x have no logic to encode a `UIAnimatedImage` into disk with animation format like GIF.

This may cost some accident effect for users. User may accidently encode GIF (which consume much more CPU && RAM) into disk, but they actually need the poster frame. So I don't know whether it's suitable to chagne or not.

**Comment by kinarobin** [comment]
It looks more like a bug. I think cache should keep its original format. 

I just debugged the gif image. If encode it in PNG format,  it will use the `SDImageAPNGCoder` to encode the gif image.  It doesn't  consume much more CPU && RAM from the perspective of encoding, just like before. Some of the differences are just encoding in JPEG format.

Even so, I still think the user should pay for his behavior. If you want store the GIF image, you should understand that this operation will consume much more CPU && RAM, otherwise you should store a frame of the GIF.

**Comment by gonghonglou** [comment]
Yes, if i store a gif to disk, then i take out it, it's shuld be a gif format. Now, i see @dreampiggy has a PR for this question.

By the way, SD can only store images downloaded by SD in the correct format, if the gif image from other way, and without data, SD still can't find the correct format. 
So, if we can get the image format with UIImage, it's perfect, but I can't find a solution...

**Comment by dreampiggy** [comment]
> It looks more like a bug. I think cache should keep its original format.
> 
> I just debugged the gif image. If encode it in PNG format, it will use the `SDImageAPNGCoder` to encode the gif image. It doesn't consume much more CPU && RAM from the perspective of encoding, just like before. Some of the differences are just encoding in JPEG format.
> 
> Even so, I still think the user should pay for his behavior. If you want store the GIF image, you should understand that this operation will consume much more CPU && RAM, otherwise you should store a frame of the GIF.

We can not have the code below:

```objective-c
if (image.sd_isAnimated) {
    // UIAnimatedImage
    data = [SDAPNGImageCoder encodedDataWithImage:image format:SDImageFormatPNG];
}
```

only can have the code like this:

```objective-c
if (image.sd_imageFormat != SDImageFormatUndefined) {
    // Any image
    data = [SDImageCodersManager encodedDataWithImage:image format:image.sd_imageFormat];
}
```

Because APNG encode will trigger the bug before macOS 10.14, see the bugreport here:

![image](https://user-images.githubusercontent.com/6919743/75641375-08643200-5c73-11ea-83f9-77ed8735019f.png)

If you want to handle the `UIAnimatedImage` (on iOS), or `NSBitmapImage with GIF`, you'd better encode it into GIF, not the APNG, although I think APNG is better than GIF.

**Comment by dreampiggy** [comment]
@gonghonglou  Solved by #2953. Should be available in v5.6.0

**Comment by gonghonglou** [comment]
> @gonghonglou Solved by #2953. Should be available in v5.6.0

got it
